src/systems/audio-zones-system.js

```diff
@@ -0,0 +1,161 @@
+
+    if (index !== -1) {
+      this.zones.splice(index, 1);
+    }
+  }
+
+  tick() {
+    if (!this.scene.is("entered") || this.zones.length === 0) return;
+
+    // Update zones
+    this._updateZones();
+
+    this.sources.forEach(source => {
+      // Only check whenever either the source or the listener have updated zones (moved)
+      if (source.entity.isUpdated() || this.listener.entity.isUpdated()) {
+        // Cast a ray from the listener to the source
+        this.rayDir = source
+          .getPosition()
+          .clone()
+          .sub(this.listener.getPosition());
+        this.normalizedRayDir = this.rayDir.clone().normalize();
+        this.ray.set(this.listener.getPosition(), this.normalizedRayDir);
+
+        // First we check the zones the source is contained in and we check the inOut property
+        // to modify the sources audio params when the listener is outside the source's zones
+        // We always apply the outmost active zone audio params, the zone that's closest to the listener
+        const inOutParams = source.entity
+          .getZones()
+          .filter(zone => {
+            const zoneBBAA = zone.getBoundingBox();
+            this.intersectTarget = new THREE.Vector3();
+            this.ray.intersectBox(zoneBBAA, this.intersectTarget);
+            return this.intersectTarget !== null && zone.data.inOut && !this.listener.entity.getZones().includes(zone);
+          })
+          .map(zone => zone.getAudioParams())
+          .reduce(paramsReducer, null);
+
+        // Then we check the zones the listener is contained in and we check the outIn property
+        // to modify the sources audio params when the source is outside the listener's zones
+        // We always apply the inmost active zone audio params, the zone that's closest to the listener
+        const outInParams = this.listener.entity
+          .getZones()
+          .filter(zone => {
+            const zoneBBAA = zone.getBoundingBox();
+            this.intersectTarget = new THREE.Vector3();
+            this.ray.intersectBox(zoneBBAA, this.intersectTarget);
+            return this.intersectTarget !== null && zone.data.outIn && !source.entity.getZones().includes(zone);
+          })
+          .map(zone => zone.getAudioParams())
+          .reduce(paramsReducer, null);
+
+        // Resolve the zones
+        if (outInParams || inOutParams) {
+          const params = outInParams ? outInParams : inOutParams;
+          params.gain = outInParams && inOutParams ? Math.min(outInParams.gain, inOutParams.gain) : params.gain;
+          params.coneOuterGain =
+            outInParams && inOutParams
+              ? Math.min(outInParams.coneOuterGain, inOutParams.coneOuterGain)
+              : params.coneOuterGain;
+          source.apply(params);
+        } else {
+          source.restore();
+        }
+      }
+    });
+  }
+
+  // Updates zones states in case they have changed.
+  _updateZones() {
```

Comment:
They are two different updates, in the zones update we iterate over the zones and then we iterate over the sources to update their states based on the zones state. Not sure if it could be merged...